From ae15bc816699920606f453dd9e134df301ed97ea Mon Sep 17 00:00:00 2001
From: Chen Wandun <chenwandun@huawei.com>
Date: Mon, 9 May 2022 20:50:19 +0800
Subject: [PATCH 2191/2592] [Huawei] psi: fix wrong iteration in iterate_groups

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I562O9

--------------------------------

It is different to get the cgroup that is used to update psi info in
cgroup v1 and cgroup v2.

task_cgroup can only used in cgroup v1, so add branch to achieve this.

Signed-off-by: Chen Wandun <chenwandun@huawei.com>
---
 kernel/sched/psi.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/kernel/sched/psi.c b/kernel/sched/psi.c
index 0c9a596692a5..c84c6f7b59fd 100644
--- a/kernel/sched/psi.c
+++ b/kernel/sched/psi.c
@@ -757,9 +757,13 @@ static struct psi_group *iterate_groups(struct task_struct *task, void **iter)
 			cgroup = task->cgroups->dfl_cgrp;
 		else {
 #ifdef CONFIG_CGROUP_CPUACCT
-			rcu_read_lock();
-			cgroup = task_cgroup(task, cpuacct_cgrp_id);
-			rcu_read_unlock();
+			if (!cgroup_subsys_on_dfl(cpuacct_cgrp_subsys)) {
+				rcu_read_lock();
+				cgroup = task_cgroup(task, cpuacct_cgrp_id);
+				rcu_read_unlock();
+			} else {
+				cgroup = task->cgroups->dfl_cgrp;
+			}
 #else
 			cgroup = NULL;
 #endif
-- 
2.17.1

